src/Channels.Networking.Sockets/SocketConnection.cs

```diff
@@ -242,104 +263,110 @@ private enum BufferStyle
-                        {
-                            throw new SocketException((int)args.SocketError);
-                        }
 
-                        // note we can't check BytesTransferred <= 0, as we always
-                        // expect 0; but if we returned, we expect data to be
-                        // buffered *on the socket*, else EOF
-                        if ((bytesFromInitialDataBuffer = args.BytesTransferred) <= 0)
+                        int bytesFromInitialDataBuffer = 0;
+
+                        if (Socket.Available == 0)
                         {
-                            if ((object)initialDataBuffer == (object)_zeroLengthBuffer)
+                            // now, this gets a bit messy unfortunately, because support for the ideal option
+                            // (zero-length reads) is platform dependent
+                            switch (_bufferStyle)
                             {
-                                // sentinel value that means we should just
-                                // consume sync (we expect there to be data)
-                                initialDataBuffer = null;
+                                case BufferStyle.Unknown:
+                                    try
+                                    {
+                                        initialSegment = await ReceiveInitialDataUnknownStrategyAsync(args);
+                                    }
+                                    catch
+                                    {
+                                        initialSegment = default(ArraySegment<byte>);
+                                    }
+                                    if (initialSegment.Array == null)
+                                    {
+                                        continue; // redo from start
+                                    }
+                                    break;
+                                case BufferStyle.UseZeroLengthBuffer:
+                                    // if we already have a buffer, use that (but: zero count); otherwise use a shared
+                                    // zero-length; this avoids constantly changing the buffer that the args use, which
+                                    // avoids some overheads
+                                    args.SetBuffer(args.Buffer ?? _zeroLengthBuffer, 0, 0);
+                                    if (Socket.ReceiveAsync(args))
+                                    {
+                                        // wait async for the io work to be completed
+                                        await ((Signal)args.UserToken).WaitAsync();
+                                    }
+                                    break;
+                                case BufferStyle.UseSmallBuffer:
+                                    // We need  to do a speculative receive with a *cheap* buffer while we wait for input; it would be *nice* if
```

Comment:
Only thing I can really do is code against the exposed API; this *is* a valid concern, however, because if that blocks `ReceiveFromSocketAndPushToChannelAsync`, it could block the `SocketConnection` constructor. The fact that  it passes the tests says it is probably not too broken.